storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+      zero.interval_ticks = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    case kYearMonthInterval: {
+      YearMonthIntervalLit zero;
+      zero.months = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    default: {
+      CHECK(false) << "SMA Index was created on an unindexable type.";
+    }
+  }
+}
+
+}  // namespace sma_internal
+
+
+SMAIndexSubBlock::SMAIndexSubBlock(const TupleStorageSubBlock &tuple_store,
+                                   const IndexSubBlockDescription &description,
+                                   const bool new_block,
+                                   void *sub_block_memory,
+                                   const std::size_t sub_block_memory_size)
+    : IndexSubBlock(tuple_store,
+                    description,
+                    new_block,
+                    sub_block_memory,
+                    sub_block_memory_size),
+      header_(nullptr),
+      entries_(nullptr),
+      total_attributes_(tuple_store.getRelation().size()),
+      attribute_to_entry_(nullptr),
+      indexed_attributes_(0),
+      initialized_(false),
+      add_operations_(new UncheckedBinaryOperator*[kNumTypeIDs]),
+      less_comparisons_(new UncheckedComparator*[kNumTypeIDs]),
+      equal_comparisons_(new UncheckedComparator*[kNumTypeIDs]) {
+  CHECK(DescriptionIsValid(relation_, description_))
+      << "Attempted to construct an SMAIndexSubBlock from an invalid description.";
+
+  indexed_attributes_ = description.ExtensionSize(SMAIndexSubBlockDescription::indexed_attribute_id);
+
+  CHECK((sizeof(sma_internal::SMAHeader)
+            + (indexed_attributes_ * sizeof(SMAEntry))) <= sub_block_memory_size_)
+      << "Attempted to create SMAIndexSubBlock without enough space allocated.";
+
+  header_ = reinterpret_cast<SMAHeader*>(sub_block_memory_);
+  entries_ = reinterpret_cast<SMAEntry*>(header_ + 1);
+  attribute_to_entry_ = new int[total_attributes_];
+  // Set each attribute to invalid.
+  for (std::size_t i = 0; i < total_attributes_; ++i) {
+    attribute_to_entry_[i] = -1;
+  }
+
+  if (new_block) {
+    header_->index_consistent = false;
+  }
+
+  // Zero the comparator arrays so we don't try to reference invalid memory later.
+  memset(less_comparisons_, 0, sizeof(UncheckedComparator*) * kNumTypeIDs);
+  memset(equal_comparisons_, 0, sizeof(UncheckedComparator*) * kNumTypeIDs);
+  memset(add_operations_, 0, sizeof(UncheckedBinaryOperator*) * kNumTypeIDs);
+
+  // Iterate through each attribute which is being indexed.
+  for (std::size_t indexed_attribute_num = 0;
+       indexed_attribute_num < indexed_attributes_;
+       ++indexed_attribute_num) {
+    const attribute_id attribute = description_.GetExtension(
+        SMAIndexSubBlockDescription::indexed_attribute_id,
+        indexed_attribute_num);
+
+    const Type &attribute_type = tuple_store_.getRelation().getAttributeById(attribute)->getType();
```

Comment:
@pateljm  Yep, and something which @hbdeshmukh explained (outside of github) was that adding these checks to areas where they might seemingly never occur protects from future, unanticipated changes.